### PR TITLE
Fix issue 13582: Don't print deprecation when importing a deprecated module from another deprecated module.

### PR DESCRIPTION
**Pull request by FeepingCreature**
There was a previous PR: https://github.com/dlang/dmd/pull/4048 but it got closed unmerged.
This PR only addresses a subset, but it's a subset that was personally annoying me and the fix is cheap, so there you go.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
If this only addresses a subset it seems to me that it shouldn't fix the issue, no?

**Comment by FeepingCreature** [comment]
Eh. The fix proposed in the bug report is kinda wrong: deprecated A imports B imports deprecated C doesn't warn in their view, but I'd argue the deprecated C import needs to warn because somebody could later import B from, say, non-deprecated D, at which point you'd need to transitively walk the module graph to figure out everything that now needs to be warned.

The more limited fix of this PR is arguably all that can practically be done.

**Comment by Geod24** [comment]
>  deprecated A imports B imports deprecated C doesn't warn in their view, but I'd argue the deprecated C import needs to warn because somebody could later import B from, say, non-deprecated D, at which point you'd need to transitively walk the module graph to figure out everything that now needs to be warned.

Completely agree, and that's how things currently work.
E.g. take the following code:
```d
deprecated struct Bar (T)
{
    T v;
}

struct Foo (T)
{
    Bar!T v;
}

deprecated struct FooBar
{
    Foo!(int) v;
}
```

Even though `Foo` in instantiated from a deprecated struct, it emits a deprecation message while using `Bar`, because `Foo`'s scope is not deprecated.

@FeepingCreature : A self import is not quite a good test IMO. Could you:
- Import another module
- Add a non deprecated module which imports locally in a deprecated aggregate (`struct`, `class`) and a deprecated function ?

**Comment by FeepingCreature** [comment]
Thanks @Geod24! The implementation as it was completely disabled import deprecation accidentally. For some reason there was no test for that, so I added one too (fail13582). Also something is wrong with `scope.isDeprecated`, and I'll look into why it doesn't do what it says it does, but for now this seems to work.

**Comment by FeepingCreature** [comment]
Update: current version expands `Scope.isDeprecated` to check the containing module as well.

**Comment by FeepingCreature** [comment]
Update: removed fail13582 because it was already being checked by fail19609. Not sure why self-imported avoided that issue but I don't think that's worth a testcase.

File: test/compilable/test13582b.d [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// REQUIRED_ARGS: -de
```
Comment (Geod24) [inline test/compilable/test13582b.d]: Output check ?

**Comment by Geod24** [comment]
@FeepingCreature : Let me know when you are happy with this and I'll test it and see if I can come up with a few new test cases.

File: test/compilable/test13582b.d (hunk 1 above)
Comment (FeepingCreature) [reply test/compilable/test13582b.d]: For what? There's not supposed to be any output, and if there's any it'd be an error and fail that way.

**Comment by FeepingCreature** [comment]
I think it's good now, but I thought that the last few times I had to hurriedly fix things too.

**Comment by Geod24** [comment]
So it looks like function-local imports don't trigger deprecation messages... Make `foo` non-deprecated and it will still not trigger a message, with your PR or with DMD 2.087.1 (and I guess with master).

**Comment by Geod24** [comment]
Also, a deprecated import inside a string `mixin` does not trigger an error (with latest release).

**Comment by FeepingCreature** [comment]
Hm... should I take a look or should those go in separate bugs?

File: test/compilable/test13582b.d (hunk 1 above)
Comment (Geod24) [reply test/compilable/test13582b.d]: Yeah I somehow didn't register they were in compilable, my bad.

**Comment by Geod24** [comment]
As far as I can tell, this PR is good to go. You added a test for function-local import so when someone actually gets around fixing this bug, the desired behavior will be enforced.
I'll open bugzilla issues for what I mentioned.

**Review by Geod24 — APPROVED** [review]

**Comment by Geod24** [comment]
For reference:
https://issues.dlang.org/show_bug.cgi?id=20163
https://issues.dlang.org/show_bug.cgi?id=20164

**Pull request merged**